<a href="https://colab.research.google.com/github/varya4414/Diabetes-Prediction/blob/master/Diabetes_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [0]:
diabetes_data=pd.read_csv("/content/drive/My Drive/Machine_learning_Diabetes/diabetes.csv")

In [4]:
print(diabetes_data.columns)


Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')


In [5]:
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1

[5 rows x 9 columns]

In [6]:
diabetes_data.shape

(768, 9)

In [7]:
diabetes_data.groupby('Outcome').size()

Outcome
0    500
1    268
dtype: int64

In [8]:
diabetes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
Pregnancies                 768 non-null int64
Glucose                     768 non-null int64
BloodPressure               768 non-null int64
SkinThickness               768 non-null int64
Insulin                     768 non-null int64
BMI                         768 non-null float64
DiabetesPedigreeFunction    768 non-null float64
Age                         768 non-null int64
Outcome                     768 non-null int64
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(diabetes_data.loc[:, diabetes_data.columns != 'Outcome'], diabetes_data['Outcome'],random_state=66)


In [10]:
X_train.shape

(576, 8)

In [11]:
y_train.shape

(576,)

In [12]:
print(X_train.columns)

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age'],
      dtype='object')


# **Decision Tree**

In [13]:
from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier(random_state=44)
tree_model.fit(X_train, y_train)


DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=44, splitter='best')

In [14]:
from sklearn import metrics
test_prediction_tree=tree_model.predict(X_test)
metrics.accuracy_score(y_test,test_prediction_tree)

0.7083333333333334

In [15]:
train_prediction_tree=tree_model.predict(X_train)
metrics.accuracy_score(y_train,train_prediction_tree)

1.0

Here when we use Decision Tree  train_accuracy = 100.0%
                          and   test_accuracy  = 64.56%
                          
    So we can say it is High Variance.
    So we fix max_depth=3.

In [16]:
improve_tree_model = DecisionTreeClassifier(max_depth=3,random_state=44)
improve_tree_model.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=3,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=44, splitter='best')

In [17]:
improve_test_prediction_tree=improve_tree_model.predict(X_test)
metrics.accuracy_score(y_test,improve_test_prediction_tree)

0.7135416666666666

In [18]:
improve_train_prediction_tree=improve_tree_model.predict(X_train)
metrics.accuracy_score(y_train,improve_train_prediction_tree)

0.7934027777777778

After fix the depth of tree to 3 we can not improve accuracy on test data.

# Random Forest

In [19]:
from sklearn.ensemble import RandomForestClassifier

forest_model = RandomForestClassifier(random_state=44)
forest_model.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/ensemble/forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=44, verbose=0,
                       warm_start=False)

In [20]:
test_prediction_forest=forest_model.predict(X_test)
metrics.accuracy_score(y_test,test_prediction_forest)

0.7552083333333334

In [21]:
train_prediction_forest=forest_model.predict(X_train)
metrics.accuracy_score(y_train,train_prediction_forest)

0.9826388888888888

So here also accuracy of train data is 98.61%.
                         and test data is 75.00%
     so there is no improvement in accuracy.

# Support vector machine

In [22]:
from sklearn.svm import SVC

model_svc = SVC(random_state=44)
model_svc.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='rbf', max_iter=-1, probability=False, random_state=44,
    shrinking=True, tol=0.001, verbose=False)

In [23]:
test_prediction_svc=model_svc.predict(X_test)
metrics.accuracy_score(y_test,test_prediction_svc)

0.6979166666666666

In [24]:
train_prediction_svc=model_svc.predict(X_train)
metrics.accuracy_score(y_train,train_prediction_svc)

1.0

model overfit with a perfect score on the training set and only 69.80% accuracy on the test set

SVM requires all the features to vary on a similar scale. We will need to re-scale our data that all the features are approximately on the same scale

In [0]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

In [26]:
scaled_model_svc = SVC(random_state=55)
scaled_model_svc.fit(X_train_scaled, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='rbf', max_iter=-1, probability=False, random_state=55,
    shrinking=True, tol=0.001, verbose=False)

In [27]:
scaled_test_prediction_svc=scaled_model_svc.predict(X_test_scaled)
metrics.accuracy_score(y_test,scaled_test_prediction_svc)

0.75

In [28]:
scaled_train_prediction_svc=scaled_model_svc.predict(X_train_scaled)
metrics.accuracy_score(y_train,scaled_train_prediction_svc)

0.8298611111111112

Now data is underfit so we can now use more complex data.
like increase value of C or gamma.

# Deep Learning

In [29]:
from sklearn.neural_network import MLPClassifier
mlp_model = MLPClassifier(random_state=44)
mlp_model.fit(X_train, y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100,), learning_rate='constant',
              learning_rate_init=0.001, max_iter=200, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=44, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [30]:
from sklearn import metrics
test_prediction_mlp=mlp_model.predict(X_test)
metrics.accuracy_score(y_test,test_prediction_mlp)

0.703125

In [31]:
train_prediction_mlp=mlp_model.predict(X_train)
metrics.accuracy_score(y_train,train_prediction_mlp)

0.7673611111111112

Here accuracy is not good. May be it is bacause scalling of data. So we re-scale the data.

In [32]:
mlp_model_scaled = MLPClassifier(random_state=44)
mlp_model_scaled.fit(X_train_scaled, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100,), learning_rate='constant',
              learning_rate_init=0.001, max_iter=200, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=44, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [33]:
test_prediction_mlp_scaled=mlp_model_scaled.predict(X_test_scaled)
metrics.accuracy_score(y_test,test_prediction_mlp_scaled)

0.7708333333333334

In [34]:
train_prediction_mlp_scaled=mlp_model_scaled.predict(X_train_scaled)
metrics.accuracy_score(y_train,train_prediction_mlp_scaled)

0.8298611111111112

Increase number of iteratotion to 1000 

In [38]:
improved_mlp_model_scaled = MLPClassifier(random_state=44,max_iter=1000)
improved_mlp_model_scaled.fit(X_train_scaled, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100,), learning_rate='constant',
              learning_rate_init=0.001, max_iter=1000, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=44, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [39]:
improved_test_prediction_mlp_scaled=improved_mlp_model_scaled.predict(X_test_scaled)
metrics.accuracy_score(y_test,improved_test_prediction_mlp_scaled)

0.7447916666666666

In [40]:
improved_train_prediction_mlp_scaled=improved_mlp_model_scaled.predict(X_train_scaled)
metrics.accuracy_score(y_train,improved_train_prediction_mlp_scaled)

0.9288194444444444

So maximum accuracy is 77.08 % which we are getting in Deep learning